docs/2019-09-09-Combining-Rewrite-Axioms.md

```diff
@@ -0,0 +1,172 @@
+   // Renaming variables to make things clear
+1. ∀ Y .
+        ( φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉
+        → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉) ∧ (⌈α(Y) ∧ φ(X)⌉)
+        )
+   // a ∧ b → c is the same as a ∧ b → c ∧ b
+```
+
+When doing normal rewriting, we usually expect to get substitutions for the
+variables in `Y` and `Y’` when computing `⌈α(Y) ∧ φ(X)⌉`, which usually allows
+us to remove these variables. However, when combining rewriting axioms,
+we don’t always get such substitutions, so we need to take a different approach.
+
+First, let us note that if `φ₁`, `φ₂` and `φ₃` are *functional*, then
+```
+⌈φ₁∧φ₂⌉ ∧ ⌈φ₁∧φ₃⌉
+= (φ₁ == φ₂) ∧ (φ₁ == φ₃)
+= (φ₁ == φ₂) ∧ (φ₂ == φ₃)
+= ⌈φ₁∧φ₂⌉ ∧ ⌈φ₂∧φ₃⌉
+```
+
+So we have
+
+```
+∀ Y .
+   ( φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉
+   → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉) ∧ (⌈α(Y) ∧ φ(X)⌉)
+   )
+iff
+∀ Y .
+   ( φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉
+   → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉) ∧ (⌈α(Y’) ∧ α(Y)⌉)
+   )
+```
+
+Now, if, additionally, `α(Y)` is *constructor-based* (or, at least,
+injection-based), then
+```
+⌈α(Y’) ∧ α(Y)⌉ \iff Y = Y’
+```
+Note that this might also hold in some constrained cases when using
+non-constructor LHS, e.g. LHS that use maps or sets.
+These are not detailed here.
+
+So, assuming that, we have the following transformations:
+```
+1. ∀ Y .
+      ( φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉
+      → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉) ∧ (⌈α(Y’) ∧ α(Y)⌉)
+      )
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉) ∧ (Y’=Y))
+   // Apply the formula above
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y’ . (•β(Y) ∧ ⌈α(Y) ∧ φ(X)⌉) ∧ (Y’=Y))
+   // Apply the substitution
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → (•β(Y) ∧ ⌈α(Y) ∧ φ(X)⌉) ∧ (∃ Y’ . Y’=Y))
+   // ∃ Y . γ ∧ φ(Y) = γ ∧ ∃ Y . φ(Y)
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → (•β(Y) ∧ ⌈α(Y) ∧ φ(X)⌉)
+   // (∃ Y . Y == γ) = ⊤
+1. ∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ⌈α(Y) ∧ φ(X)⌉ ∧ (•β(Y))
+   // s(γ ∧ P) = s(γ) ∧ P if s is a symbol and P is a predicate
+1. ∀ Y . ((φ(X) → •β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉)
+   // (a ∧ c) → (b ∧ c)
+1. (φ(X) → •β(Y)) ∧ ⌈α(Y) ∧ φ(X)⌉
+   // ∀ is not needed at the top level
+```
+
+Note that if `⌈α(Y) ∧ φ(X)⌉` contains a substitution `y=γ` then we can apply it
+and remove the variable from the expression above (proof not shown here, one
+needs to transform
+```
+∀ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉ → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉))).
+```
+into
+```
+∃ Y . (φ(X) ∧ ⌈α(Y) ∧ φ(X)⌉) → ∃ Y’ . (•β(Y’) ∧ ⌈α(Y’) ∧ φ(X)⌉)
+```
+above to make it work).
+
+Combining rewrite rules
+=======================
+
+Let’s say that our axioms are `α₁(X₁)⇒β₁(X₁)` … `αₙ(Xₙ)⇒βₙ(Xₙ)`.
+
+Let us attempt to combine `α₁(X₁)⇒β₁(X₁)` with `α₂(X₂)⇒β₂(X₂)`. Let us assume
+that, when applying `α₂(X₂)⇒β₂(X₂)` to `β₁(X₁)` as described above, we get
+```
+(β₁(X₁) → •β₂(X₂)) ∧ ⌈β₁(X₁)∧α₂(X₂)⌉
+```
+
+Note that
+```
+P ∧ (a→•b) = (P∧ a) → •P ∧ b
+```
+And, if we also know that `P ∧ b→• P ∧ c`, then we can infer
+`(P∧ a) → ••(P ∧ c)`, i.e. `P ∧ (a→••c)`.
+
+TODO: Does the above hold? Why?
```

Comment:
it holds due to transitivity of the implication and due to the "Framing rule". Say you have `a →• b` and `b →• c`
From `b →• c` you deduce by framing  `• b →• • c`, and then from transitivity of implication and MP the conclusion follows (modulo the extraction of P which we now know it's not correct).